### PR TITLE
Use replication controller asynchronous API

### PR DESCRIPTION
**Pull request by fjammes**
- Upgrade qserv-operator
- Remove use of multithread for ingest
- Wait for operator startup in CI
- Improve load balancer on contribution files
- Rename 'chunk file' to 'contribution' in source code
- Add failed load balancer for dataservers in integration tests
- Improve code style
- Add unit test for empty DNS entry
- Fix requests retry object
- Refine http time-out management
- Remove ingested contribution from ingest loop,
  which start ingest and monitor non-ingested contributions
- Add interactive step to workflow

### CODE REVIEW

File: doc/user/run.rst [hunk 1]
```diff
@@ -97,8 +97,23 @@ Launch the workflow using `Argo <https://argoproj.github.io/argo/>`__
...
+=================================
+
+
+.. code:: sh
+
+    ./argo-submit.sh -s
+    # Retrive the pod name for the 'interactive' step
+    argo get @latest
+    # Open a shell inside it
+    kubectl exec -it qserv-ingest-2hrcf-595146013 -c main bash
```
Comment (iagaponenko) [inline doc/user/run.rst:117]: I'm a specific name of a pod mentioned in the command. Should it be replaced with some variable? Or it doesn't matter here since this is just an example?

File: manifests/workflow.yaml [hunk 2]
```diff
@@ -82,6 +82,13 @@ spec:
         template: ingest-step
         arguments:
           parameters: [{name: script, value: benchmark.sh}]
+  # Allow to run interactively some ingest workflow steps
+  - name: interactive
+    steps:
+    - - name: interactive
```
Comment (iagaponenko) [inline manifests/workflow.yaml:88]: Two `- -` here. Is this an array of arrays?

File: rootfs/ingest/python/qserv/ingest.py [hunk 3]
```diff
@@ -62,175 +55,181 @@
...
-        then load it inside Qserv,
+        Get contributions from a queue server for a given database
+        then ingest it inside Qserv,
         during a super-transation
 
         Returns:
         --------
-        Integer number: 0 if no more chunk to load,
-                        1 if a chunk was loaded successfully
+        Integer number: 0 if no more contribution to load,
```
Comment (iagaponenko) [inline rootfs/ingest/python/qserv/ingest.py:147]: Perhaps:
```
        Integer number: 0 if no more contributions to load,
```

File: rootfs/ingest/python/qserv/ingest.py [hunk 4]
```diff
@@ -62,175 +55,181 @@
...
+                        1 if at least a contribution was loaded successfully
```
Comment (iagaponenko) [inline rootfs/ingest/python/qserv/ingest.py]: Perhaps:
```
                        1 if at least one contribution was loaded successfully
```

File: rootfs/ingest/python/qserv/ingest.py [hunk 5]
```diff
@@ -62,175 +55,181 @@
...
             return False
 
         transaction_id = None
-        chunk = None
         try:
-            transaction_id = self.repl_client.start_transaction(self.chunk_meta.database)
-            ingest_success = self._ingest_parallel_contributions(transaction_id, chunks_locked)
+            transaction_id = self.repl_client.start_transaction(self.contribution_metadata.database)
+
+            contributions = build_contributions(contributions_locked,
```
Comment (iagaponenko) [inline rootfs/ingest/python/qserv/ingest.py:161]: I think this operation should be done outside (before) this `try ... except` block and before starting the transaction. A problem is that exception handling is not designed to process exceptions thrown by the method `build_contributions`. As a result of that, any problems encountered within the method `build_contributions` would lead to strange errors and recovery actions.

File: rootfs/ingest/python/qserv/ingest.py [hunk 6]
```diff
@@ -62,175 +55,181 @@
...
+            for i, c in enumerate(contributions):
+                # Ingest to start
+                if c.request_id is None:
+                    # Ingest to start
+                    startedAt = time.strftime("%H:%M:%S", time.localtime())
+                    _LOG.debug('Contribution %s ingest started at %s', c, startedAt)
+                    c.start_async(transaction_id)
+                    idx_contribs_in_progress.append(i)
+                else:
+                    startedAt = time.strftime("%H:%M:%S", time.localtime())
```
Comment (iagaponenko) [inline rootfs/ingest/python/qserv/ingest.py]: Code duplication since you're already initializing to exactly the same value in the `True` section of the `if` block. You may push both lines above `if` as:
```
startedAt = time.strftime("%H:%M:%S", time.localtime())
if c.request_id is None:
    LOG.debug('Contribution %s ingest started at %s', c, startedAt)
    ....
else:
    _LOG.debug('Contribution %s ingest monitored at %s', c, startedAt)
    ....
```

File: rootfs/ingest/python/qserv/ingest.py [hunk 7]
```diff
@@ -62,175 +55,181 @@
...
+                    _LOG.debug('Contribution %s ingest monitored at %s', c, startedAt)
+                    if c.monitor():
```
Comment (iagaponenko) [inline rootfs/ingest/python/qserv/ingest.py]: Will this "monitor" also check failures in ingesting a contribution? Or, perhaps, the "monitor" throws an exception if it detects any such problem? I'm just not seeing where such (failures) conditions are being handled in the current method.

File: rootfs/ingest/python/qserv/ingest.py [hunk 8]
```diff
@@ -62,175 +55,181 @@
...
+                contribs_in_progress.append(contributions[i])
+            contributions = contribs_in_progress
+
+            contribs_in_progress_count = len(contributions)
+            if contribs_in_progress_count == 0:
+                finished = True
+            if not finished:
+                _LOG.debug('Processing %s contributions for transaction %s',
+                           contribs_in_progress_count, transaction_id)
+                time.sleep(5)
```
Comment (iagaponenko) [inline rootfs/ingest/python/qserv/ingest.py:219]: I'm not seeing how this `while not finished:` loop will end if the worker ingests service marks some contributions as failed. My impression is that the loop will go forever since it only handles successful contributions.

File: rootfs/ingest/python/qserv/ingest.py [hunk 9]
```diff
@@ -62,175 +55,181 @@
...
+                    # Ingest to start
+                    startedAt = time.strftime("%H:%M:%S", time.localtime())
+                    _LOG.debug('Contribution %s ingest started at %s', c, startedAt)
+                    c.start_async(transaction_id)
+                    idx_contribs_in_progress.append(i)
+                else:
+                    startedAt = time.strftime("%H:%M:%S", time.localtime())
+                    _LOG.debug('Contribution %s ingest monitored at %s', c, startedAt)
+                    if c.monitor():
+                        # Ingest successful
```
Comment (iagaponenko) [inline rootfs/ingest/python/qserv/ingest.py]: Only "successful"? Or maybe "completed"? Note that "completed" includes "successfully ingested" and "failed to be ingested" as reported by the worker ingest service.

File: rootfs/ingest/python/qserv/jsonparser.py [hunk 10]
```diff
@@ -64,13 +78,36 @@ def filter_transactions(responseJson, database, states):
...
-                        trans['id'], trans['state'])
+                       trans['id'], trans['state'])
             state = TransactionState(trans['state'])
             if state in states:
                 transaction_ids.append(trans['id'])
     return transaction_ids
 
-def get_indexes(responseJson, existing_indexes: typing.Dict[str, set]=dict()):
+
+def get_contribution_status(responseJson: dict) -> ContributionState:
```
Comment (iagaponenko) [inline rootfs/ingest/python/qserv/jsonparser.py:88]: In the spirit of the OO programming, I would suggest moving this method into the enumeration class:
```
class ContributionState(Enum):
    CANCELLED = "CANCELLED"
    ..
    LOAD_FAILED = "LOAD_FAILED"

    @classmethod
    def parse(status: str) -> ContributionState:
        if status == cls.FINISHED.value: return cls.FINISHED
        elif status == cls.IN_PROGRESS.value: return cls.IN_PROGRESS
        ...
        else:
            raise ReplicationControllerError(f"Unknown contribution status: {status}")
```
After that, you would use the embedded method like this:
```
  state = ContributionState.parse(responseJson['contrib']['status'])
```
Putting together data members with operations over the data provided a better abstraction for the class and avoid mistakes since it would be easier to track states.

File: rootfs/ingest/python/qserv/test_loadbalancerurl.py [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+def test_get_loadbalancer_url():
+    path = "/lsst-dm/qserv-ingest/master/tests/data/cosmoDC2/"
+    servers = [
+        "https://server1",
+        "https://server2",
+        "https://server3"
+    ]
+    lb_url = LoadBalancedURL(path, servers)
+    url = lb_url.get()
+    assert (url == "https://server1/lsst-dm/qserv-ingest/master/tests/data/cosmoDC2/")
```
Comment (iagaponenko) [inline rootfs/ingest/python/qserv/test_loadbalancerurl.py]: Is it always `https:`? It looks like this scheme is hard-coded into the implementation of the balancer.

File: rootfs/ingest/python/qserv/test_loadbalancerurl.py [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+    filename = "file.txt"
+    servers = [
+        "https://server1",
+        "https://server2",
+        "https://server3"
+    ]
+    lb_url = LoadBalancedURL(path, servers)
+    lb_url.join(filename)
+    url = lb_url.get()
+    assert (url == f"https://server1{path}{filename}")
```
Comment (iagaponenko) [inline rootfs/ingest/python/qserv/test_loadbalancerurl.py]: GitHub complains - no newline in the end of the file.

File: rootfs/ingest/python/qserv/contribution.py [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+                                    load_balanced_base_url)
+        contributions.append(contribution)
+    return contributions
+
+
+class Contribution():
+    """Represent an ingest contribution
+    """
+
+    def __init__(self, worker_host, worker_port, chunk_id, path, table, is_overlap,
```
Comment (iagaponenko) [inline rootfs/ingest/python/qserv/contribution.py]: Don't you want to do type annotations for the input parameters?
```
    def __init__(self, worker_host: str, worker_port: int, chunk_id: int, path: str, table: str, is_overlap: bool,
```

File: rootfs/ingest/python/qserv/http.py [hunk 14]
```diff
@@ -88,24 +94,26 @@ def json_get(base_url, filename):
...
             return json.load(f)
     else:
-        raise Exception("Unsupported URI scheme for ", url)
+        raise IngestError("Unsupported URI scheme for ", url)
+
 
-class TimeoutHTTPAdapter(HTTPAdapter):
-    """ Manage http connection time-out
-        See https://findwork.dev/blog/advanced-usage-python-requests-timeouts-retries-hooks/
+def _get_retry_object(retries=5, backoff_factor=0.2):
```
Comment (iagaponenko) [inline rootfs/ingest/python/qserv/http.py]: Type annotation for the returned type?
```
def _get_retry_object(retries=5, backoff_factor=0.2) -> Retry:
```

**Review by iagaponenko — APPROVED** [review]
To the best of my knowledge of Python, looks okay to me. I still have a few general comments on this PR. First of all, it would be nice if it was also reviewed by one of the DAX team's Python experts.  Secondly, the commit structure made it very hard to review the changes because most of the changes were due to renaming `chunk` into `contribution`, etc. Ideally, trivial refactoring like that is better be done in a separate (the very first) commit on the PR.

File: rootfs/ingest/bin/pytest.sh [hunk 15]
```diff
@@ -2,5 +2,4 @@
 
 # run unit tests inside qserv-ingest container
 pip install pytest
-pytest /ingest/python/ --log-cli-level=DEBUG
-
+pytest /ingest/python/ --log-cli-level=DEBUG --capture=tee-sys
```
Comment (andy-slac) [inline rootfs/ingest/bin/pytest.sh]: `log_cli_level` is also defined in `pytest.ini`, may be better to keep all options in one place?

File: rootfs/ingest/python/qserv/contribution.py [hunk 16]
```diff
@@ -0,0 +1,181 @@
+#!/usr/bin/env python
+
+# LSST Data Management System
+# Copyright 2014-2015 AURA/LSST.
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/contribution.py]: Year should be changed, and I think it's not AURA

File: rootfs/ingest/python/qserv/contribution.py [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+                        load_balanced_base_url: LoadBalancedURL) -> list:
+    """Build list of contributions to be ingested
+
+    Args:
+        contributions_locked (_type_): _description_
+        repl_client (_type_): _description_
+        load_balanced_url (list): _description_
+
+    Returns:
+        list: contributions to be ingested
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/contribution.py]: Docstrings should be formatted according to LSST style guide: https://developer.lsst.io/python/numpydoc.html

File: rootfs/ingest/python/qserv/contribution.py [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+
+class Contribution():
+    """Represent an ingest contribution
+    """
+
+    def __init__(self, worker_host, worker_port, chunk_id, path, table, is_overlap,
+                 load_balanced_base_url):
+        """Store input parameters for replication REST api located at
+           http://<worker_host:worker_port>/ingest/file-async
+        """
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/contribution.py]: `__init__` docstring is not used by LSST, documentation of constructor arguments and what it does is a part of class docstring.

File: rootfs/ingest/python/qserv/contribution.py [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+    def __str__(self):
+        return f"Contribution({self.__dict__})"
+
+    def start_async(self, transaction_id: int) -> None:
+        """Start an asynchronous ingest query for a chunk contribution
+           Raise an exception if the query fails after a fixed number of attempts
+           (see MAX_RETRY_ATTEMPTS constant)
+
+        Args:
+            transaction_id (int): id of the transaction in which the contribution will be ingested
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/contribution.py:119]: This line looks longer than what LSST recommends. If we stick with LSST guidelines it may be worth running `flake8` on Python sources (and mypy/black/isort if you feel adventurous).

File: rootfs/ingest/python/qserv/contribution.py [hunk 20]
```diff
@@ -0,0 +1,181 @@
...
+        """Monitor an asynchronous ingest query for a chunk contribution
+           States of an ingest query are listed here:
+           https://confluence.lsstcorp.org/display/DM/Ingest%3A+9.5.3.+Asynchronous+Protocol
+           Propagate an exception if the query fails after a fixed number of attempts
+
+        Raises:
+            Exception: in case of error during contribution ingest
+
+        Returns:
+            bool: True if contribution has been successfully ingested
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/contribution.py]: Order of Raises and Returns should be opposite.

File: rootfs/ingest/python/qserv/contribution.py [hunk 21]
```diff
@@ -0,0 +1,181 @@
...
+        """
+        status_url = urllib.parse.urljoin(self.worker_url,
+                                          f"ingest/file-async/{self.request_id}")
+        responseJson = Http().get(status_url)
+
+        contrib_status = get_contribution_status(responseJson)
+        match contrib_status:
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/contribution.py:173]: Do we always guarantee that this runs with Python 3.10?

File: rootfs/ingest/python/qserv/contribution.py [hunk 22]
```diff
@@ -0,0 +1,181 @@
...
+                                      f"Contribution {self}")
+                self.retry_attempts += 1
+                self.request_id = None
+            case _:
+                retry = raise_error(responseJson, self.retry_attempts, MAX_RETRY_ATTEMPTS)
+                if retry:
+                    self.retry_attempts += 1
+                    self.request_id = None
+                else:
+                    raise ReplicationControllerError(f"Contribution {self}s should be in an error" +
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/contribution.py]: `s` after `{self}` looks like typo.

File: rootfs/ingest/python/qserv/exception.py [hunk 23]
```diff
@@ -0,0 +1,39 @@
+#!/usr/bin/env python
+
+# LSST Data Management System
+# Copyright 2014-2015 AURA/LSST.
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/exception.py]: year/AURA

File: rootfs/ingest/python/qserv/http.py [hunk 24]
```diff
@@ -114,85 +122,70 @@ class Http():
...
-        adapter = TimeoutHTTPAdapter(max_retries=retry_strategy)
+        adapter = HTTPAdapter(max_retries=_get_retry_object())
+        # Session is only used for the GET method
         self.http = requests.Session()
         self.http.mount("https://", adapter)
         self.http.mount("http://", adapter)
 
-    def get(self, url, payload=dict(), auth=True) -> dict:
-        if auth == True:
+    def get(self, url, payload=dict(), auth=True, timeout=None) -> dict:
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/http.py:131]: Missing docstring (and other methods below).

File: rootfs/ingest/python/qserv/http.py [hunk 25]
```diff
@@ -114,85 +122,70 @@ class Http():
...
+            r = requests.post(url, json=payload, timeout=timeout)
         except (requests.exceptions.RequestException, ConnectionResetError) as e:
             _LOG.critical("Error when sending POST request to url %s", url)
             e.args = (f"POST request to url {url} with payload {payload} failed", *e.args)
             raise e
         if (r.status_code != 200):
-            raise Exception(
-                'Error in HTTP response (POST)', url,
-                r.status_code)
+            raise HTTPError('HTTP %s Error for url (POST): %s' % (r.status_code, url), response=r)
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/http.py:158]: f-string for consistency?

File: rootfs/ingest/python/qserv/ingest.py [hunk 26]
```diff
@@ -62,175 +55,181 @@
...
 
-    def _ingest_parallel_contributions(self, transaction_id: int, chunks_locked):
-        args = self._compute_args(chunks_locked, transaction_id)
-        with ThreadPool() as pool:
-            for chunk, startedAt, endedAt in pool.imap_unordered(_ingest_contribution, args):
-                logging.debug('Chunk %s ingest started at %s ended at %s' % (chunk, startedAt, endedAt))
+    def _ingest_all_contributions(self, transaction_id: int, contributions: list) -> bool:
+        """Ingest all contribution for a given transaction
+           Throw exception if ingest fail
+           This method always returns True, or raise an exception
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/ingest.py:182]: If it always returns true, it probably does not make sense to return it at all?

File: rootfs/ingest/python/qserv/jsonparser.py [hunk 27]
```diff
@@ -64,13 +78,36 @@ def filter_transactions(responseJson, database, states):
...
+    elif responseJson['contrib']['status'] == ContributionState.CANCELLED.value:
+        return ContributionState.CANCELLED
+    elif responseJson['contrib']['status'] == ContributionState.CREATE_FAILED.value:
+        return ContributionState.CREATE_FAILED
+    elif responseJson['contrib']['status'] == ContributionState.START_FAILED.value:
+        return ContributionState.START_FAILED
+    elif responseJson['contrib']['status'] == ContributionState.READ_FAILED.value:
+        return ContributionState.READ_FAILED
+    elif responseJson['contrib']['status'] == ContributionState.LOAD_FAILED.value:
+        return ContributionState.LOAD_FAILED
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/jsonparser.py:104]: You don't need the long if/else list,this should work:

    try:
        return ContributionState(responseJson['contrib']['status'])
    except ValueError:
        raise ReplicationControllerError("Unknown contribution status:" +
                                         f"{responseJson['contrib']['status']}")

And I agree with Igor, better make it a class method

File: rootfs/ingest/python/qserv/loadbalancerurl.py [hunk 28]
```diff
@@ -0,0 +1,92 @@
+#!/usr/bin/env python
+
+# LSST Data Management System
+# Copyright 2014-2015 AURA/LSST.
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/loadbalancerurl.py]: Year/AURA

File: rootfs/ingest/python/qserv/loadbalancerurl.py [hunk 29]
```diff
@@ -0,0 +1,92 @@
...
+# Imports for other modules --
+# ----------------------------
+
+# ---------------------------------
+# Local non-exported definitions --
+# ---------------------------------
+_LOG = logging.getLogger(__name__)
+
+
+class LoadBalancedURL():
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/loadbalancerurl.py:46]: `()` not needed after class name.

File: rootfs/ingest/python/qserv/loadbalancerurl.py [hunk 30]
```diff
@@ -0,0 +1,92 @@
...
+
+        Args:
+            path (str): path of the url
+            loadbalancers (List[str], optional): List of http(s) load balancer urls. Defaults to [].
+
+        Raises:
+            ValueError: if path uses an unsupported protocol
+        """
+
+        if len(loadbalancers) != 0:
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/loadbalancerurl.py:63]: A usual Pythonic expression is:

    if loadbancers:
        ...

File: rootfs/ingest/python/qserv/loadbalancerurl.py [hunk 31]
```diff
@@ -0,0 +1,92 @@
...
+        self.loadbalancers = []
+        if url.scheme in ["http", "https"]:
+            self.loadbalancers = loadbalancers
+        elif url.scheme != "file":
+            raise ValueError("Unsupported scheme for URL: %s, %s", path, loadbalancers)
+
+    def __repr__(self):
+        return f"LoadBalancedURL({self.__dict__})"
+
+    def get(self) -> str:
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/loadbalancerurl.py:79]: docstrings are missing for methods.

File: rootfs/ingest/python/qserv/replicationclient.py [hunk 32]
```diff
@@ -175,25 +172,37 @@ def database_register_tables(self, tables_json_data, felis):
...
         url = urllib.parse.urljoin(self.repl_url, "replication/config")
-        responseJson = self.http.get(url)
+        responseJson = self.http.get(url, timeout=self.timeout_short)
         jsonparser.raise_error(responseJson)
         status = jsonparser.parse_database_status(responseJson, database, family)
         _LOG.debug(f"Database {family}:{database} status: {status}")
         return status
 
+    @lru_cache(maxsize=None)
+    def get_chunk_location(self, chunk, database):
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/replicationclient.py]: docstring is missing

File: rootfs/ingest/python/qserv/test_contribution.py [hunk 33]
```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/env python
+
+# LSST Data Management System
+# Copyright 2014-2015 AURA/LSST.
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/test_contribution.py]: year

File: rootfs/ingest/python/qserv/test_contribution.py [hunk 34]
```diff
@@ -0,0 +1,89 @@
...
+           "is_overlap": True,
+           "load_balanced_base_url": LB_URL}
+
+
+def test_init():
+    params = _PARAMS
+    contribution = Contribution(**params)
+
+    url = "https://server{}/lsst/data/step1_1/chunk_1_overlap.txt"
+    assert(contribution.load_balanced_url.get() == url.format(1))
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/test_contribution.py]: `assert` is not a function, it does not need parentheses.

File: rootfs/ingest/python/qserv/test_contribution.py [hunk 35]
```diff
@@ -0,0 +1,89 @@
...
+
+def test_print():
+
+    c = Contribution(**_PARAMS)
+    print(c)
+    _LOG.debug(c)
+
+    params = _PARAMS
+
+    params.pop('load_balanced_base_url')
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/test_contribution.py]: Double quotes should be used in most cases for consistency.

File: rootfs/ingest/python/qserv/test_loadbalancerurl.py [hunk 36]
```diff
@@ -0,0 +1,70 @@
+#!/usr/bin/env python
+
+# LSST Data Management System
+# Copyright 2014-2015 AURA/LSST.
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/test_loadbalancerurl.py]: year

**Review by andy-slac — APPROVED** [review]
Looks OK, my comments are mostly about formatting for docstrings and comments, would be nice to keep that consistent with LSST style guide.

**Review by fritzm — APPROVED** [review]
Looks good to me, given comments already provided by Igor and Andy

File: doc/user/run.rst (hunk 1 above)
Comment (fjammes) [reply doc/user/run.rst:117]: Yes it is just an example.

File: manifests/workflow.yaml (hunk 2 above)
Comment (fjammes) [reply manifests/workflow.yaml:88]: yes

File: rootfs/ingest/python/qserv/ingest.py (hunk 4 above)
Comment (fjammes) [reply rootfs/ingest/python/qserv/ingest.py]: :+1:

File: rootfs/ingest/python/qserv/ingest.py (hunk 5 above)
Comment (fjammes) [reply rootfs/ingest/python/qserv/ingest.py:161]: the method `build_contribution` connect to the replication service in order to get contributions location. Be aware that the code catches any exception and will abort the transaction if it catches one.

File: rootfs/ingest/python/qserv/ingest.py (hunk 6 above)
Comment (fjammes) [reply rootfs/ingest/python/qserv/ingest.py]: :+1:

File: rootfs/ingest/python/qserv/ingest.py (hunk 7 above)
Comment (fjammes) [reply rootfs/ingest/python/qserv/ingest.py]:
```
    def monitor(self) -> bool:
        """Monitor an asynchronous ingest query for a chunk contribution
           States of an ingest query are listed here:
           https://confluence.lsstcorp.org/display/DM/Ingest%3A+9.5.3.+Asynchronous+Protocol
           Propagate an exception if the query fails after a fixed number of attempts

        Raises:
            Exception: in case of error during contribution ingest

        Returns:
            bool: True if contribution has been successfully ingested
                  False if contribution is being ingested
        """
```

File: rootfs/ingest/python/qserv/ingest.py (hunk 8 above)
Comment (fjammes) [reply rootfs/ingest/python/qserv/ingest.py:219]: An unsuccessful contribution will spawn an exception and exit the loop

Comment (fjammes) [reply rootfs/ingest/python/qserv/ingest.py:219]: I have tried to simplify the algorithm.

File: rootfs/ingest/python/qserv/ingest.py (hunk 9 above)
Comment (fjammes) [reply rootfs/ingest/python/qserv/ingest.py]: It is FINISHED (i.e. "successfully ingested"), fixed :+1:

File: rootfs/ingest/python/qserv/test_loadbalancerurl.py (hunk 11 above)
Comment (fjammes) [reply rootfs/ingest/python/qserv/test_loadbalancerurl.py]: No, can can set the protocol inside the `servers` variable.

File: rootfs/ingest/python/qserv/test_loadbalancerurl.py (hunk 12 above)
Comment (fjammes) [reply rootfs/ingest/python/qserv/test_loadbalancerurl.py]: :+1:

File: rootfs/ingest/python/qserv/contribution.py (hunk 13 above)
Comment (fjammes) [reply rootfs/ingest/python/qserv/contribution.py]: :+1:

File: rootfs/ingest/python/qserv/http.py (hunk 14 above)
Comment (fjammes) [reply rootfs/ingest/python/qserv/http.py]: :+1:

File: rootfs/ingest/bin/pytest.sh (hunk 15 above)
Comment (fjammes) [reply rootfs/ingest/bin/pytest.sh]: :+1: done

File: rootfs/ingest/python/qserv/contribution.py (hunk 16 above)
Comment (fjammes) [reply rootfs/ingest/python/qserv/contribution.py]: :+1: done

File: rootfs/ingest/python/qserv/contribution.py (hunk 17 above)
Comment (fjammes) [reply rootfs/ingest/python/qserv/contribution.py]: :+1:

File: rootfs/ingest/python/qserv/contribution.py (hunk 18 above)
Comment (fjammes) [reply rootfs/ingest/python/qserv/contribution.py]: :+1:

File: rootfs/ingest/python/qserv/contribution.py (hunk 19 above)
Comment (fjammes) [reply rootfs/ingest/python/qserv/contribution.py:119]: :+1:

File: rootfs/ingest/python/qserv/replicationclient.py (hunk 32 above)
Comment (fjammes) [reply rootfs/ingest/python/qserv/replicationclient.py]: :+1:

File: rootfs/ingest/python/qserv/test_contribution.py (hunk 33 above)
Comment (fjammes) [reply rootfs/ingest/python/qserv/test_contribution.py]: :+1:

File: rootfs/ingest/python/qserv/test_contribution.py (hunk 34 above)
Comment (fjammes) [reply rootfs/ingest/python/qserv/test_contribution.py]: :+1:

File: rootfs/ingest/python/qserv/test_loadbalancerurl.py (hunk 36 above)
Comment (fjammes) [reply rootfs/ingest/python/qserv/test_loadbalancerurl.py]: :+1:

File: rootfs/ingest/python/qserv/test_contribution.py (hunk 35 above)
Comment (fjammes) [reply rootfs/ingest/python/qserv/test_contribution.py]: :+1:
